kubernetes/README.md

```diff
@@ -15,5 +15,30 @@ After the snapshot is created in a temporary directory, `s3cmd` is used to sync
 * `S3_URI` - S3 URI to use to upload (s3://xxx)
 * `S3_BUCKET` - S3 bucket to point to
 * `S3_HOST` - S3 endpoint
+* `S3_EXPIRE_DAYS` - Delete files older than this threshold (expired)
 * `AWS_ACCESS_KEY_ID` - Access key to use to access S3
 * `AWS_SECRET_ACCESS_KEY` - Secret access key to use to access S3
+
+## Configuration of file retention (pruning)
+
+With AWS S3, use [lifecycle
+rules](https://docs.aws.amazon.com/AmazonS3/latest/userguide/lifecycle-expire-general-considerations.html)
+to configure retention and automatic cleanup action (prune) for expired files.
+
+For other S3 compatible storage, ensure to set [Governance
+lock](https://community.exoscale.com/documentation/storage/versioning/#set-up-the-lock-configuration-for-a-bucket)
+to avoid any modification before `$S3_EXPIRE_DAYS`:
+
+```
+mc retention set --default GOVERNANCE "${S3_EXPIRE_DAYS}d" my-s3-remote/my-bucket
+```
+
+On removal by the `vault-snapshot.sh` script, [`DEL` deletion marker
```

Comment:
@4censord thank you for your help, I put the command to see the deletion markers in the docs here, because that behavior is a bit confusing :smile: 

Also, it's not clear to me yet if a (Governance) "locked" object that was deleted (`DEL`) will automatically be pruned at the end of the lock/retention time, or if it waits until the next `DEL` command. Maybe consecutive `DEL` also don't have any effect, depending on how that is implemented.. let's test

Reply:
@4censord we need to discuss again what the Governance lock does exactly:
* I can have a governance lock with retention of 1 day
* I delete a snapshot from 2024-09-06 (>1 day ago)
* The tombstone (`DEL` marker) is set

So this is the exact same behavior like when I configure a Governance lock with retention of 90 days. I'm confused :confused:

Reply:
I think I found my answer.

The behavior is indeed the same. You can delete and "undo" the tombstones (`DEL` markers) as many times as you wish, irrespective of the Governance lock.

It will only be different once you try to touch the actual snapshot files. These are protected by the lock (visible in the metadata):
```
Metadata  :
  X-Amz-Object-Lock-Mode             : GOVERNANCE 
  X-Amz-Object-Lock-Retain-Until-Date: 2024-12-11T21:41:28.753Z 
  Content-Type                       : application/binary 
```

So the files itself cannot be deleted, not even with `mc rm --versions --force`. You can "bypass" governance retention with the correct permissions. However, I'm not really sure that is implemented correctly (I tested on Exoscale). I was unable to do for instance this:

```bash
mc rm --bypass my-snapshots/vault-snapshots/vault_2024-09-12-2341.snapshot
mc: <ERROR> Failed to remove `my-snapshots/vault-snapshots/vault_2024-09-12-2341.snapshot`. You are not authorized to perform bypass-governance-retention on bucket vault-snapshots
```

Reply:
on exoscale bypassing the governance lock is a special permission that afaik can not be assigned with just policies, but basically an extra checkmark while creating a set of credentials

Reply:
Slightly related - I'm currently also observing a weird behavior on cloudscale LPG, where the Governance lock is "reset" a few seconds after being set. Really weird behavior, because it seems to be working in the other region RMA. I opened a support request to clarify..